### ISSUE TITLE
BFS, DOBFS and SSSP report incorrect results for random and RMAT-style graphs

### ISSUE THREAD

**Issue by sree314** [gunrock/gunrock#63]
I modified these programs (patch included) to store the results in an
output file which is then compared against a database of reference
outputs we maintain (created using Merrill's BFS and Davidson's SSSP).

I enclose the smallest rmat graph where the incorrect behaviour
manifests along with the reference outputs and the output generated by
gunrock.

Notably, the number of unreachable nodes is higher in gunrock compared
to the reference outputs:

`$ grep --count '2147483646' gunrock-bfs-rmat10-0.txt bfs-rmat10-0-ref.txt
gunrock-bfs-rmat10-0.txt:108
bfs-rmat10-0-ref.txt:97`

The same is true for SSSP.

As an example of incorrect weights, the rmat10 graph contains an SSSP
path of weight 1493 that is not detected by gunrock:

`$ grep '1493' gunrock-sssp-rmat10-0.txt sssp-rmat10-0-ref.txt
sssp-rmat10-0-ref.txt:16 1493
sssp-rmat10-0-ref.txt:945 1493`

The DOBFS also seems to have the same issues.

I have observed similar errors on random graphs, but not on road networks.

Interestingly, in all cases, the CPU and GPU agree!

[GitHub did not allow me to attach the files, and I've emailed them to you]

**Comment by yzhwang** [comment]
I think I know where the confusion came from. All inconsistencies are due to the different graph IO implementation. In Gunrock we read in each line of matrix market format in column-major order, meaning:
col, row, val
https://github.com/gunrock/gunrock/blob/master/gunrock/graphio/market.cuh#L160
This causes all our directed graph results different from the reference algorithm if in reference algorithm the IO is implemented in the other way. We didn't notice this because when we did our performance comparison, we always added --undirected to form the undirected graph.

Will fix this once I make sure this is the only reason that causes the problem. Good catch, thanks!

**Comment by sree314** [comment]
Yep, I fixed that and the results for DOBFS, BFS and SSSP now match the reference outputs.

I guess that also explains why the results for road networks were correct

I checked the formats for Matrix Market and there seems to be no column-major order. The clearest statement that it is always in row, column, value is here:

http://people.sc.fsu.edu/~jburkardt/data/mm/mm.html

Thank you for the quick response!
